### ISSUE TITLE
False positive E1121 when using multiple inheritance and overriding class names

### ISSUE THREAD

**Issue by iantra** [pylint-dev/pylint#9087]
### Bug description

```python
Originally found this via VSCode while using the google cloud API, tried my best to turn it into a minimal example.

If we have a file `classes.py`:
 python
class A:
    def func(self, arg1, arg2):
        super().func([arg1, arg2])


class B:
    def func(self, args):
        print(args)


class B(A, B):
    pass
```

and a second file `weird.py`:
``` python
from classes import B

obj = B()
obj.func(1, 2)
```
we get an unexpected E1121 on line 4.

The file runs without any errors, and this doesn't happen when the classes are declared in the same file that calls `func`, or if the final class has a different name (e.g. "C"). Also doesn't seem to happen with single inheritance (e.g. declaring a class "A" that subclasses the original "A" and overwrites `func`)

Possibly related to https://github.com/pylint-dev/pylint/issues/1123 ?
```


### Configuration

_No response_

### Command used

```shell
pylint weird.py
```

### Pylint output

```python
************* Module weird
weird.py:1:0: C0114: Missing module docstring (missing-module-docstring)
weird.py:4:0: E1121: Too many positional arguments for method call (too-many-function-args)
```

### Expected behavior

I would expect no E1121, since the program runs through and we (correctly) get no E1121 when we move the class definition into the main file

### Pylint version

```shell
pylint 2.17.7
astroid 2.15.8
Python 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0]
```

**Comment by nickdrozd** [comment]
Simpler reproducing case with trace printing added:

```python
class A:
    def func(self, arg1, arg2):
        print(f'A.func({arg1}, {arg2})')
        super().func([arg1, arg2])


class B:
    def func(self, args):
        print(f'B.func({args})')


class C(A, B):
    pass


C().func(1, 2)
```

Here is the output of running this:

```
A.func(1, 2)
B.func([1, 2])
```

`C` has no `func` method defined, so `C.func` delegates to `A.func`. That method in turn calls `super().func`. I would have that this would call the superclass of `A` (i.e. `object`), but no; instead, it calls the next superclass of `C`, namely `B`. And that's how `B.func` gets called.

This behavior is surprising to me. What exactly does `super()` do?

**Comment by nickdrozd** [comment]
Mypy doesn't like the `super` call from `A`: `error: "func" undefined in superclass  [misc]`.

This seems to be an example of code that happens to work out at runtime but that is difficult to analyze statically. Unless there is an easy fix to deal with this, I think it's okay if Pylint rejects it.

**Comment by iantra** [comment]
> Simpler reproducing case with trace printing added:
> 
> ```python
> class A:
>     def func(self, arg1, arg2):
>         print(f'A.func({arg1}, {arg2})')
>         super().func([arg1, arg2])
> 
> 
> class B:
>     def func(self, args):
>         print(f'B.func({args})')
> 
> 
> class C(A, B):
>     pass
> 
> 
> C().func(1, 2)
> ```
> 
> Here is the output of running this:
> 
> ```
> A.func(1, 2)
> B.func([1, 2])
> ```
> 
> `C` has no `func` method defined, so `C.func` delegates to `A.func`. That method in turn calls `super().func`. I would have that this would call the superclass of `A` (i.e. `object`), but no; instead, it calls the next superclass of `C`, namely `B`. And that's how `B.func` gets called.
> 
> This behavior is surprising to me. What exactly does `super()` do?

Maybe I'm missing something, but this doesn't seem to reproduce the E1121 error (too many positional arguments) on my end. All I get is the "func" undefined error you mentioned, which I'm also ok with given that it **is** in fact undefined unless A is used in this very specific way (multiple inheritance). To be clear, my issue is only with the E1121 error (that only shows up in the file that uses the classes), not any of the other ones that show up in the actual class declarations, since those all make sense to me and don't really affect me as the end-user of a library that happens to do weird things under the hood.

As for the whole super() issue, that's just how python deals with multiple inheritance (by following the [MRO](https://www.geeksforgeeks.org/method-resolution-order-in-python-inheritance/), which for this situation is (B, A, B, object) and you can check with a print(B.__mro__)). From what I can tell this is the intended behavior for the code and multiple inheritance in general, not some weird accident.

Not sure about how difficult it would be to analyze, but given that the error doesn't show up when the classes are defined in the same file that calls the function, it seems like the MRO is already being taken into account and resolved correctly in other situations, so I can't imagine it would be too hard to solve it for this case. Funny enough, it seems mypy does show an error in that case (everything in one file) which means pylint is already doing something extra to take the MRO into account, it just somehow doesn't extend into things imported from other files properly.

**Comment by nickdrozd** [comment]
Oh I see, this is about `too-many-function-args`. Well, Mypy raises the same warning in the same place: `error: Too many arguments for "func" of "B"  [call-arg]`. So I'm inclined to say that this code is just hard to understand.

Specifically, `class B(A, B)` seems like a strange thing to do. I assume you have a compelling reason for doing it, and I'm interested to hear why. But in most cases I think that it would be an error that the linter should flag.

Let's say instead you use `class C(A, B)` to avoid the name clash, but you want to leave all of your existing `B` callsites undisturbed. Then in `weird.py` you could change the import to use an alias:

```python
from classes import C as B

obj = B()
obj.func(1, 2)
```

Both Pylint and Mypy accept that version of the code.

So there are a variety of code changes that could be made to avoid the warning. Barring that, the warning can disabled either locally or globally. But that's about it. I'm skeptical that Pylint should attempt to accommodate this circumstance.
